src/data_source/fetching.rs

```diff
@@ -93,21 +253,36 @@ where
     S: NodeDataSource<Types>
         + UpdateNodeData<Types>
         + AvailabilityStorage<Types>
-        + VersionedDataSource,
-    P: Send + Sync,
+        + VersionedDataSource
+        + 'static,
+    P: AvailabilityProvider<Types>,
 {
-    /// Create a data source with local storage and a remote data availability provider.
-    pub async fn new(storage: S, provider: P) -> anyhow::Result<Self> {
-        Self::with_retry_delay(storage, provider, None).await
+    /// Build a [`FetchingDataSource`] with the given `storage` and `provider`.
+    pub fn builder(storage: S, provider: P) -> Builder<Types, S, P> {
+        Builder::new(storage, provider)
     }
 
-    pub async fn with_retry_delay(
-        storage: S,
-        provider: P,
-        delay: Option<Duration>,
-    ) -> anyhow::Result<Self> {
+    /// Create a data source with local storage and a remote data availability provider.
+    pub async fn new(builder: Builder<Types, S, P>) -> anyhow::Result<Self> {
```

Comment:
Not 100% following...are you suggesting basically replace `FetchingDataSource::new` with `Builder::build` or something? I wanted to keep the implementation of this function in `FetchingDataSource` since the logic is specific to this type. But interface wise I agree...maybe we should make this function private and have `Builder::build` be the only way to construct a `FetchingDataSource`. WDYT?

Reply:
Essentially yes.  Following a Traditional `Builder` pattern the `build` call should do all of the setup work.  But at the same time the `FetchingDataSource::new` call really shouldn't depend on the `Builder`, as it implies it is unable to be built without a `Builder`, which essentially makes it not the `Builder` pattern.


As an experiment, I wrote the following types to try and work around it:

```rust
pub struct FetchingDataSourceConfig<Types, S, P>
where
    Types: NodeType,
{
    minor_scan_interval: Duration,
    major_scan_interval: usize,
    proactive_fetching: bool,
    fetcher: Fetcher<Types, S, P>,
}
```

Then the `FecthingDataSource::new` function would look something similar to:

```rust
pub async fn new(config: FetchingDataSourceConfig<Types, S, P>) -> anyhow::Result<Self> {
        let proactive_fetching = config.proactive_fetching;
        let minor_interval = config.minor_scan_interval;
        let major_interval = config.major_scan_interval;

        let fetcher = Arc::new(config.fetcher);
        let scanner = if proactive_fetching {
            Some(BackgroundTask::spawn(
                "proactive scanner",
                fetcher
                    .clone()
                    .proactive_scan(minor_interval, major_interval),
            ))
        } else {
            None
        };

        let mut ds = Self {
            fetcher,
            scanner,
            metrics: Default::default(),
        };

        if NodeDataSource::block_height(&ds).await? == 0 {
            // HotShot doesn't emit an event for the genesis block, so we need to manually ensure it
            // is present.
            ds.insert_genesis().await?;
        }

        Ok(ds)
    }
```

But we'd also need to do the same for the `Fetcher`, since it has the same dependency in it's `Fetcher::new` call.

```rust
pub struct FetcherConfig<Types, S, P>
where
    Types: NodeType,
{
    storage: S,
    provider: P,
    retry_delay: Option<Duration>,
    range_chunk_size: usize,

    _types: PhantomData<Types>,
}
```

Thus making the `Fetcher::new` call look like:

```
pub struct FetcherConfig<Types, S, P>
where
    Types: NodeType,
{
    storage: S,
    provider: P,
    retry_delay: Option<Duration>,
    range_chunk_size: usize,

    _types: PhantomData<Types>,
}
```

And finally resulting in a `Builder::build` function that looks like this:

```rust
pub async fn build(self) -> anyhow::Result<FetchingDataSource<Types, S, P>> {
        let fetcher = Fetcher::new(FetcherConfig {
            storage: self.storage,
            provider: self.provider,
            retry_delay: self.retry_delay,
            range_chunk_size: self.range_chunk_size,
            _types: Default::default(),
        })
        .await?;

        FetchingDataSource::new(FetchingDataSourceConfig {
            minor_scan_interval: self.minor_scan_interval,
            major_scan_interval: self.major_scan_interval,
            proactive_fetching: self.proactive_fetching,
            fetcher: fetcher,
        })
        .await
    }
```

This removes the `Builder` from being an argument to the `Fetcher::new`, and `FetchingDataSource::new` calls while still allowing the `Builder::build` to create a new instance of `FetchingDataSource`.  In addition it keeps the building portion of the `Builder::build` call all internalized as much as possible.  It also allows `Fetcher::new` and `FetchingDataSource::new` to be created without a `Builder`, should one choose to.

The down side is two new bespoke configuration types for each.  The alternative would be to keep adding more arguments to the `new` functions, but that makes any changes to their data type requirements more annoying to change at the function invocation site.  Either way the Call site would need to suffer, without the `Builder`.

Reply:
Functionally there's no difference here.  Things just get moved around.  But a `Builder` as a concept remains `distinct`.  You *could* achieve the same affect by renaming the `new` functions, and keeping them scoped so that only the `Builder` can see them.

Reply:
Gotcha. That all makes sense. For the time being I don't think we gain much from being able to construct the `FetchingDataSource` without the builder via `new`, and as you pointed at there's a maintenance burden from introducing those new config types. So for now my inclination is to make the unnecessary constructors private, and we can improve the interface if/when we ever have a need to expose them.

Reply:
Done in [a663b2c](https://github.com/EspressoSystems/hotshot-query-service/pull/390/commits/a663b2c7dad400dcbeb80d6ceb2da11580f24b70)